contracts/FundDistribution.sol

```diff
@@ -0,0 +1,257 @@
+  mapping(address => mapping(address => uint256)) public tokenAvailable;
+  //all the current tokens in the contract
+  address[] public tokens;
+  mapping(address => bool) public distributors;
+  mapping(address => bool) public funders;
+
+  event EthApproveIsSet(address to, uint256 amount);
+  event TokenApproveIsSet(address to, address token, uint256 amount);
+  event NewTokenAdded(address sender, address token, uint256 balance);
+  event TokenIsAdded(address sender, address token, uint256 amount);
+  event EthIsAdded(address sender, uint256 amount);
+  event FundIsClaimed(address to);
+  event EthIsClaimed(address to, uint256 amount);
+  event TokenIsClaimed(address to, address token, uint256 amount);
+  event AllTokensAreClaimed(address to);
+  event ClaimedPartial(address to, address token, uint256 amount);
+  event ClaimedEthPartial(address to, uint256 amount);
+
+  constructor(address _owner) public {
+    owner = _owner;
+  }
+
+  modifier onlyNonZeroAmount(uint256 amount) {
+    require(amount > 0, "Invalid amount");
+    _;
+  }
+
+  modifier onlyNonZeroAddress(address to) {
+    require(to != address(0), "Invalid address");
+    _;
+  }
+
+  modifier onlyFunders() {
+    require(funders[msg.sender] || msg.sender == owner, "Only funders can call this function");
+    _;
+  }
+
+  modifier onlyDistributors() {
+    require(
+      distributors[msg.sender] || msg.sender == owner,
+      "Only distributors can call this function"
+    );
+    _;
+  }
+
+  function addFunder(address _funder) public onlyOwner {
+    funders[_funder] = true;
+  }
+
+  function addDistributor(address _distributor) public onlyOwner {
+    distributors[_distributor] = true;
+  }
+
+  receive() external payable onlyFunders onlyNonZeroAmount(msg.value) {
+    emit EthIsAdded(msg.sender, msg.value);
+  }
+
+  //don't emit an event as receive() already emitted it
+  function depositEth() public payable onlyFunders onlyNonZeroAmount(msg.value) {}
+
+  function depositEthAndApprove(address to) public payable {
+    depositEth();
+    setEthApprove(to, msg.value);
+  }
+
+  //transfer token to this contract first and then call this function
+  function addTokenAfterTransfer(address token) public onlyFunders onlyNonZeroAddress(token) {
+    require(!curTokens[token], "Token already added");
+    curTokens[token] = true;
+    tokens.push(token);
+    uint256 amount = IERC20(token).balanceOf(address(this));
+    require(amount > 0, "Token has no balance");
+    emit TokenIsAdded(msg.sender, token, amount);
+  }
+
+  // approve allowance first then call receiveToken to transfer token
+  function depositToken(address token, uint256 amount)
+    public
+    onlyFunders
+    onlyNonZeroAddress(token)
+    onlyNonZeroAmount(amount)
+  {
+    if (!curTokens[token]) {
+      tokens.push(token);
+      curTokens[token] = true;
+    }
+    IERC20 tokenContract = IERC20(token);
+    tokenContract.transferFrom(msg.sender, address(this), amount);
```

Comment:
does transferFrom here work without an approve?

Reply:
No, it requires an approval before calling this function, that's what other interns also did =)).
I also provided other way to first transfer token to the contract and then add the token to the contract state later